codegen/src/lib.rs

```diff
@@ -0,0 +1,736 @@
+/// the `#[resource]` attribute.
+///
+/// ```
+/// # use legion_codegen::system;
+/// # use legion::Schedule;
+/// # struct Person { name: String }
+/// #[system]
+/// fn hello_world(#[resource] person: &Person) {
+///    println!("hello, {}", person.name);
+/// }
+/// ```
+///
+/// Systems can also request a world or command buffer.
+///
+/// ```
+/// # use legion_codegen::system;
+/// # use legion::{Schedule, systems::CommandBuffer, world::SubWorld};
+/// # struct Person { name: &'static str }
+/// #[system]
+/// fn create_entity(world: &mut SubWorld, cmd: &mut CommandBuffer) {
+///    cmd.push((1usize, false, Person { name: "Jane Doe" }));
+/// }
+/// ```
+///
+/// Systems can declare access to component types with the `#[read_component]` and
+/// `#[write_component]` attributes.
+///
+/// ```
+/// # use legion_codegen::system;
+/// # use legion::{Schedule, world::SubWorld, Read, Write, IntoQuery};
+/// # struct Time;
+/// #[system]
+/// #[read_component(usize)]
+/// #[write_component(bool)]
+/// fn run_query(world: &mut SubWorld) {
+///     let mut query = <(Read<usize>, Write<bool>)>::query();
+///     for (a, b) in query.iter_mut(world) {
+///         println!("{} {}", a, b);
+///     }
+/// }
+/// ```
+///
+/// `for_each` and `par_for_each` system types can be used to implement the query for you.
+/// References will be interpreted as `Read<T>` and `Write<T>`, while options of references
+/// (e.g. `Option<&Position>`) will be interpreted as `TryRead<T>` and `TryWrite<T>`. You can
+/// request the entity ID via a `&Entity` parameter.
+///
+/// ```
+/// # use legion_codegen::system;
+/// # struct Position { x: f32 }
+/// # struct Velocity { x: f32 }
+/// # struct Time { seconds: f32 }
+/// #[system(for_each)]
+/// fn update_positions(pos: &mut Position, vel: &Velocity, #[resource] time: &Time) {
+///     pos.x += vel.x * time.seconds;
+/// }
+/// ```
+///
+/// `for_each` and `par_for_each` systems can request attitional filters for their query via the
+/// `#[filter]` attribute.
+///
+/// ```
+/// # use legion_codegen::system;
+/// # use legion::maybe_changed;
+/// # struct Position { x: f32 }
+/// # struct Velocity { x: f32 }
+/// # struct Time { seconds: f32 }
+/// #[system(for_each)]
+/// #[filter(maybe_changed::<Position>())]
+/// fn update_positions(pos: &mut Position, vel: &Velocity, #[resource] time: &Time) {
+///     pos.x += vel.x * time.seconds;
+/// }
+/// ```
+///
+/// Systems can contain their own state. Add a reference marked with the `#[state]` parameter to
+/// your function. This state will be initialized when you construct the system.
+///
+/// ```
+/// # use legion_codegen::system;
+/// # use legion::Schedule;
+/// #[system]
+/// fn stateful(#[state] counter: &mut usize) {
+///     *counter += 1;
+///     println!("state: {}", counter);
+/// }
+///
+/// Schedule::builder()
+///      // initialize state when you construct the system
+///     .add_system(stateful_system(5usize))
```

Comment:
```suggestion
///     .add_system(stateful_system(&usize))
```

Reply:
We are initializing system state here, so we need to provide a value rather than declare a type. In this case, we initialize the system's counter to 5.

Reply:
Ah, sorry. I assumed it was a typo since I've never encountered a usize literal before. Does it not work if you just pass `5`?

Reply:
Ah, sorry. I assumed it was a typo since I've never seen a usize literal before. Does it not work if you pass `5`?

Reply:
You can suffix an numerical type to a numerical literal to explicitly have it be that type. I find it cleaner to separate it with an underscore, though, ie `5_usize`